test/typeinference/Tests/InferProgramSpec.hs

```diff
@@ -77,3 +87,112 @@ spec = do
         typedFully = lowerAndInfer fully
         typedMinimal = lowerAndInfer minimal
     typedFully `shouldBe` typedMinimal
+
+  {- A list of tricky functions to verify the correctness of type inference.
+
+  For now, I hardcode all the expected results. We should fix that later.
+  Uncomment the print statement to output the inference result for each function.
+  -}
+  it "infers id function" $ do
+    let minimal = parseProgram $ unlines
+          [ "id(x) ="
+          , "  x"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "TBuiltin (Arrow (TVar t0) (TVar t0))"
+
+  it "infers addOne function v1" $ do
+    let minimal = parseProgram $ unlines
+          [ "addOne(x) ="
+          , "  x + 1"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "TBuiltin (Arrow (TBuiltin (Integral 32)) (TBuiltin (Integral 32)))"
+
+  it "infers addOne function v2" $ do
+    let minimal = parseProgram $ unlines
+          [ "id(x) ="
+          , "  x"
+          , "addOne(x) ="
+          , "  id(x) + 1"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "TBuiltin (Arrow (TBuiltin (Integral 32)) (TBuiltin (Integral 32)))"
+
+  it "infers tricky program #1:`f` should has type `a -> a`" $ do
+    let minimal = parseProgram $ unlines
+          [ "f x ="
+          , "  let y = x"
+          , "  y"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "TBuiltin (Arrow (TVar t0) (TVar t0))"
+
+  it "infers tricky program #2:`f` should has type `a -> (b -> a)`" $ do
+    let minimal = parseProgram $ unlines
+          [ "f x ="
+          , "  let g z = x"
+          , "  g"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "Var x (TVar t0)"
+    typedPString `shouldContain` "Var g (TBuiltin (Arrow (TVar t2) (TVar t0)))"
+    typedPString `shouldContain` "TBuiltin (Arrow (TVar t0) (TBuiltin (Arrow (TVar t2) (TVar t0))))"
+
+  it "infers tricky program #3:`f` should has type `a -> (b -> b)` not `a -> (a -> a)`" $ do
+    let minimal = parseProgram $ unlines
+          [ "f x ="
+          , "  let g x = x"
+          , "  g g"
+          ]
+        typedMinimal = lowerAndInfer minimal
+        typedPString = show typedMinimal
+    -- print typedMinimal
+    typedPString `shouldContain` "Var g (TBuiltin (Arrow (TVar t4) (TVar t4)))"
```

Comment:
Fishing for the index of `TVar`s seems really fragile. I think we should write a custom comparison method for IR trees that considers identifiers modulo alpha renaming.